nighthawk/test/benchmark_http_client_test.cc

```diff
@@ -0,0 +1,289 @@
+#include "nighthawk/source/common/platform_util_impl.h"
+#include "nighthawk/source/common/rate_limiter_impl.h"
+#include "nighthawk/source/common/sequencer_impl.h"
+#include "nighthawk/source/common/statistic_impl.h"
+
+#include "envoy/thread_local/thread_local.h"
+#include "test/mocks/runtime/mocks.h"
+#include "test/mocks/thread_local/mocks.h"
+
+#include "test/integration/http_integration.h"
+
+#include "test/integration/integration.h"
+#include "test/integration/utility.h"
+#include "test/server/utility.h"
+#include "test/test_common/utility.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+
+std::string lorem_ipsum_config;
+
+class BenchmarkClientTest : public Envoy::BaseIntegrationTest,
+                            public testing::TestWithParam<Envoy::Network::Address::IpVersion> {
+public:
+  BenchmarkClientTest()
+      : Envoy::BaseIntegrationTest(GetParam(), realTime(), lorem_ipsum_config),
+        api_(thread_factory_, store_, timeSystem()), dispatcher_(api_.allocateDispatcher()) {}
+
+  static void SetUpTestCase() {
+    Envoy::Filesystem::InstanceImpl filesystem;
+
+    Envoy::TestEnvironment::setEnvVar("TEST_TMPDIR", Envoy::TestEnvironment::temporaryDirectory(),
+                                      1);
+
+    const std::string lorem_ipsum_content = filesystem.fileReadToEnd(
+        Envoy::TestEnvironment::runfilesPath("nighthawk/test/test_data/lorem_ipsum.txt"));
+    Envoy::TestEnvironment::writeStringToFileForTest("lorem_ipsum.txt", lorem_ipsum_content);
+
+    Envoy::TestEnvironment::exec({Envoy::TestEnvironment::runfilesPath("nighthawk/test/certs.sh")});
+
+    lorem_ipsum_config = filesystem.fileReadToEnd(Envoy::TestEnvironment::runfilesPath(
+        "nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml"));
+    lorem_ipsum_config = Envoy::TestEnvironment::substitute(lorem_ipsum_config);
+  }
+
+  void SetUp() override {
+    ares_library_init(ARES_LIB_INIT_ALL);
+    Envoy::Event::Libevent::Global::initialize();
+    BaseIntegrationTest::initialize();
+  }
+
+  uint32_t getTestServerHostAndPort() { return lookupPort("listener_0"); }
+
+  uint32_t getTestServerHostAndSslPort() { return lookupPort("listener_1"); }
+
+  void TearDown() override {
+    client_->terminate();
+    test_server_.reset();
+    fake_upstreams_.clear();
+    tls_.shutdownGlobalThreading();
+    ares_library_cleanup();
+  }
+
+  void setupBenchmarkClient(const std::string uriPath, bool use_https, bool use_h2,
+                            uint32_t port = 0) {
+    if (port == 0) {
+      port = use_https ? getTestServerHostAndSslPort() : getTestServerHostAndPort();
+    }
+
+    client_ = std::make_unique<Client::BenchmarkClientHttpImpl>(
+        api_, *dispatcher_, std::make_unique<Envoy::Stats::IsolatedStoreImpl>(),
+        std::make_unique<StreamingStatistic>(), std::make_unique<StreamingStatistic>(),
+        fmt::format("{}://127.0.0.1:{}{}", use_https ? "https" : "http", port, uriPath), use_h2);
```

Comment:
Nit: this will break in IPv6-only environments such as Google. This will become an issue during the Google import. See how this is handled safely with `TEST_P` and IpVersions in Envoy's `test/integration` tests.

Reply:
This is fixed, I think, but I'm looking for a way to ensure/check we used `ipv4` / `ipv6` in test expectations.